### PR TITLE
Implemented highgui module mouse wheel callback support in Linux

### PR DESCRIPTION
**Pull request by theg4sh**
The changes made possible to make custom handler on mouse wheel
events in QT and GTK for Linux opencv build.

I suppose, this feature was forgotten during implementation.

### CODE REVIEW

**Comment by StevenPuttemans** [comment]
> I suppose, this feature was forgotten during implementation.

I guess this was introduced when most mouses out there did not have a wheel :)

File: modules/highgui/src/window_gtk.cpp [hunk 1]
```diff
@@ -1973,7 +1982,25 @@ static gboolean icvOnMouse( GtkWidget *widget, GdkEvent *event, gpointer user_da
...
+            int flags = 0;
+            if (state & GDK_SCROLL_MASK) {
+                cv_event = CV_EVENT_MOUSEMOVE;
+                int orient = CV_EVENT_MOUSEWHEEL;
+                switch(event->scroll.direction) {
+                case GDK_SCROLL_RIGHT: orient = CV_EVENT_MOUSEHWHEEL;
+                case GDK_SCROLL_UP:    flags |= ((1 << 16) | orient);
+                    break;
+                case GDK_SCROLL_LEFT:  orient = CV_EVENT_MOUSEHWHEEL;
+                case GDK_SCROLL_DOWN:  flags |= ((-1 << 16) | orient);
```
Comment (alalek) [inline modules/highgui/src/window_gtk.cpp]: Looks like there is different units for CV_GET_WHEEL_DELTA macro in Win32 and GTK backends.
Win32 backend uses GET_WHEEL_DELTA_WPARAM which is "expressed in multiples or divisions of WHEEL_DELTA, which is 120".

Do you know what units is used in [gdk-event-get-scroll-deltas](https://developer.gnome.org/gdk3/stable/gdk3-Events.html#gdk-event-get-scroll-deltas) call?

File: modules/highgui/src/window_gtk.cpp [hunk 2]
```diff
@@ -1973,7 +1982,25 @@ static gboolean icvOnMouse( GtkWidget *widget, GdkEvent *event, gpointer user_da
...
+                    break;
+#if (GTK_MAJOR_VERSION > 3 or (GTK_MAJOR_VERSION == 3 and GTK_MINOR_VERSION >= 4))
```
Comment (alalek) [inline modules/highgui/src/window_gtk.cpp]: For compatibility reason we prefer to use "||" instead of "or" in conditions.

Comment (theg4sh) [inline modules/highgui/src/window_gtk.cpp]: @alalek, sure.
Due mix of programming languages in my head and the night was born this staff :)
I'll fix it soon.

File: modules/highgui/src/window_gtk.cpp (hunk 1 above)
Comment (theg4sh) [inline modules/highgui/src/window_gtk.cpp]: Nice question.. Well, the structure GdkEventScroll (event->scroll) contains delta_x and delta_y, which can be received from gdk-event-get-scroll-deltas call.
https://developer.gnome.org/gdk3/3.4/gdk3-Event-Structures.html#GdkEventScroll
I'll play with deltas and improve the code.

_I suppose, in the future CvMouseCallback can be remastered to easier access to the data of mouse event, e.g by replacing arguments 'int event, int x, int y, int flags' to 'const CvMouseEvent *'._

**Comment by theg4sh** [comment]
Seems something goes wrong with opencv-pushbot... Built was passed without errors.

**Pull request closed without merging**